File: examples/verbal_expressions/verbal_expressions.js
var _ = require('lodash');

function VerbalExpression() {
  this.raw_source = '';
}

VerbalExpression.prototype.compile = function () {
  return new RegExp(this.raw_source);
}

VerbalExpression.prototype.start_of_line = function () {
  this.raw_source += '^';
  return this;
}

VerbalExpression.prototype.maybe = function (letter) {
  this.raw_source += '(' + _.escapeRegExp(letter) + ')?';
  return this;
}

VerbalExpression.prototype.find = function (word) {
  this.raw_source += '(' + _.escapeRegExp(word) + ')';
  return this;
}

VerbalExpression.prototype.anything_but = function (letter) {
  this.raw_source += '[^' + _.escapeRegExp(letter) + ']*';
  return this;
}

VerbalExpression.prototype.end_of_line = function () {
  this.raw_source += '$';
  return this;
}

VerbalExpression.prototype.match = function (word) {
  return this.compile().exec(word);
}

VerbalExpression.prototype.source = function () {
  return this.raw_source;
}

var v = new VerbalExpression();
var a = v.start_of_line().find('http').maybe('s').find('://').maybe('www.').anything_but(' ').end_of_line();
var test_url = 'https://www.google.com';
if (a.match(test_url)) {
  console.log('Valid URL');
} else {
  console.log('Invalid URL');
}

console.log(a.source());

File: examples/verbal_expressions/verbal_expressions.go
package main

import (
	"fmt"
	"regexp"
)

type VerbalExpression struct {
	RawSource string
}

func newVerbalExpression() *VerbalExpression {
	return &VerbalExpression{""}
}

func (this *VerbalExpression) Compile() *regexp.Regexp {
	return regexp.MustCompile(this.rawSource)
}

func (this *VerbalExpression) StartOfLine() *VerbalExpression {
	this.rawSource += "^"
	return this
}

func (this *VerbalExpression) Maybe(letter string) *VerbalExpression {
	this.rawSource += fmt.Sprintf("(%s)?", regexp.QuoteMeta(letter))
	return this
}

func (this *VerbalExpression) Find(word string) *VerbalExpression {
	this.rawSource += fmt.Sprintf("(%s)", regexp.QuoteMeta(word))
	return this
}

func (this *VerbalExpression) AnythingBut(letter string) *VerbalExpression {
	this.rawSource += fmt.Sprintf("[^%s]*", regexp.QuoteMeta(letter))
	return this
}

func (this *VerbalExpression) EndOfLine() *VerbalExpression {
	this.rawSource += "$"
	return this
}

func (this *VerbalExpression) Match(word string) [][][]byte {
	return this.Compile().FindAllSubmatch([]byte(word), -1)
}

func (this *VerbalExpression) Source() string {
	return this.rawSource
}

func main() {
	v := newVerbalExpression()
	a := v.StartOfLine().Find("http").Maybe("s").Find("://").Maybe("www.").AnythingBut(" ").EndOfLine()
	testUrl := "https://www.google.com"
	if a.Match(testUrl) != nil {
		fmt.Println("Valid URL")
	} else {
		fmt.Println("Invalid URL")
	}

	fmt.Println(a.Source())
}
